### PR TITLE
Fix the worker contract.

### PR DESCRIPTION
**Pull request by restingbull**
Restructure the worker for improved (hopefully) clarity and better management of empty pipes.
Adds BazelWorkerTest that checks the behaviours of both invocation workers and persistent.

Fixed: #325.

### CODE REVIEW

File: src/main/kotlin/io/bazel/kotlin/builder/tasks/BazelWorker.kt [hunk 1]
```diff
@@ -56,107 +63,164 @@ interface CommandLineProgram {
...
-          else BazelWorker.ERROR_STATUS.also {
-            System.err.println(
-                "ERROR: Worker threw uncaught exception with args: ${args}"
-            )
-            e.printStackTrace(System.err)
+      val (status, exit) = WorkingDirectoryContext.newContext()
+        .runCatching {
+          request.argumentsList
+            ?.let {
+              maybeExpand(it)
```
Comment (cgruber) [inline src/main/kotlin/io/bazel/kotlin/builder/tasks/BazelWorker.kt:167]: nit: I'd make these one-liners. But this isn't wrong.

File: src/main/kotlin/io/bazel/kotlin/builder/tasks/BazelWorker.kt [hunk 2]
```diff
@@ -56,107 +63,164 @@ interface CommandLineProgram {
...
+                       e)
+            Status.ERROR to 1
+          } else {
+            Status.INTERRUPTED to 1
           }
-        } to buffer.toString()
+        }
+        .getOrThrow()
+
+      val response = with(WorkResponse.newBuilder()) {
```
Comment (cgruber) [inline src/main/kotlin/io/bazel/kotlin/builder/tasks/BazelWorker.kt:186]: Nit: With is fine, but also
```
val response = WorkResponse.newBuilder().apply {
  output = String(io.execution.toByteArray(), UTF_8)
  exitCode = exit
  setRequestId(request.requestId)
}.build()
```

It's a more common pattern for working with java builders, though it's the same thing.

File: src/main/kotlin/io/bazel/kotlin/builder/tasks/BazelWorker.kt [hunk 3]
```diff
@@ -56,107 +63,164 @@ interface CommandLineProgram {
...
+class InvocationWorker(
+  private val io: WorkerIO,
+  private val program: CommandLineProgram
+) : Worker {
+  val logger: Logger = Logger.getLogger(InvocationWorker::class.java.canonicalName)
+  override fun run(args: List<String>): Int {
+    return WorkingDirectoryContext.newContext().runCatching {
+      program.apply(dir, maybeExpand(args))
+    }.recover { e ->
+      logger.log(Level.SEVERE,
```
Comment (cgruber) [inline src/main/kotlin/io/bazel/kotlin/builder/tasks/BazelWorker.kt:217]: Nit. static import Level and that can be a one-liner.

Makes it a wee bit easier to see what's happening with the return code. Also, is this a place where oen of those `return@`'s will work?

File: src/main/kotlin/io/bazel/kotlin/builder/tasks/BazelWorker.kt [hunk 4]
```diff
@@ -56,107 +63,164 @@ interface CommandLineProgram {
...
-          } catch (e: IOException) {
-            throw RuntimeException(e)
-          }
-        }
+/** Defines the common worker interface. */
+interface Worker {
+  fun run(args: List<String>): Int
+}
+
+class WorkerIO(
```
Comment (cgruber) [inline src/main/kotlin/io/bazel/kotlin/builder/tasks/BazelWorker.kt:108]: This whole design is so much cleaner, thank you.

File: src/main/kotlin/io/bazel/kotlin/builder/tasks/BazelWorker.kt (hunk 1 above)
Comment (restingbull) [reply src/main/kotlin/io/bazel/kotlin/builder/tasks/BazelWorker.kt:167]: I get curly bracket blindness when the density is too tight.

File: src/main/kotlin/io/bazel/kotlin/builder/tasks/BazelWorker.kt (hunk 2 above)
Comment (restingbull) [reply src/main/kotlin/io/bazel/kotlin/builder/tasks/BazelWorker.kt:186]: Ah, cargo-culted the old code. Much cleaner, as with requires using the builder function on the last line.

File: src/main/kotlin/io/bazel/kotlin/builder/tasks/BazelWorker.kt (hunk 3 above)
Comment (restingbull) [reply src/main/kotlin/io/bazel/kotlin/builder/tasks/BazelWorker.kt:217]: Still too long for one line, I'm afraid.

**Pull request merged**